grin/src/AbstractInterpretation/EffectTracking/CodeGen.hs

```diff
@@ -0,0 +1,124 @@
+{-# LANGUAGE LambdaCase, RecordWildCards, TupleSections, TemplateHaskell, OverloadedStrings #-}
+module AbstractInterpretation.EffectTracking.CodeGen where
+
+import Control.Monad.State
+
+import Data.Set (Set)
+import Data.Map (Map)
+import Data.Vector (Vector)
+
+import qualified Data.Set as Set
+import qualified Data.Map as Map
+import qualified Data.Vector as Vec
+import Data.Functor.Foldable as Foldable
+import Data.Tuple
+import Data.Maybe
+
+import Lens.Micro.Platform
+
+import Grin.Grin
+import Grin.Pretty
+import Grin.TypeEnv
+import qualified AbstractInterpretation.IR as IR
+import AbstractInterpretation.IR (Instruction(..), AbstractProgram(..), AbstractMapping(..))
+import AbstractInterpretation.EffectTracking.CodeGenBase
+
+-- cannot return Z anywhere, any computation can contain side effects
+returnNewReg :: CG Result
+returnNewReg = do
```

Comment:
Applicative style `R <$> newReg`

Reply:
Couldn't be this a local function for the `codeGenM` if you don't intend to use anywhere else?

Reply:
Done.